File: course-scheduler.py
# Time Complexity: O(V+E)
# Space Complexity: O(n)
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        map = {}
        indegrees = [0 for z in range(numCourses)]
        for edge in prerequisites:
            indegrees[edge[0]] += 1
            if not map.has_key(edge[1]):
                map[edge[1]] = []
            map[edge[1]].append(edge[0])
        
        count = 0
        q = []
        for i in range(numCourses):
            if indegrees[i] == 0:
                q.append(i)
                count += 1
        
        if not q: return False
        while q:
            curr = q.pop()
            children = map[curr]
            if children:
                for child in children:
                    indegrees[child] -= 1
                    if indegrees[child] == 0:
                        q.append(child)
                        count += 1
        if count == 0:
            return True
        return False

File: course_scheduler.java
//Time - O(n)
//Space - O(n)
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        HashMap<Integer, ArrayList<Integer>> hm = new HashMap<>();
        Queue<Integer> q = new LinkedList<>();
        int[] courses = new int[numCourses];
        for(int[] pre: prerequisites){
            courses[pre[0]]++;
        }
        for(int[] pre: prerequisites){
            if(hm.containsKey(pre[1])){
                ArrayList<Integer> l = hm.get(pre[1]);
                l.add(pre[0]);
                hm.put(pre[1],l);
            }
            else{
                ArrayList<Integer> l = new ArrayList<>();
                l.add(pre[0]);
                hm.put(pre[1],l);

            }
        }
        for(int i=0;i<courses.length;i++){
            if(courses[i]==0){
                q.add(i);
            }
        }
        while(!q.isEmpty()){
            int n = q.size();
            for(int i=0;i<n;i++){
                int c = q.poll();
                if(hm.containsKey(c)){
                    List<Integer> sub = hm.get(c);
                    for(int j=0;j<sub.size();j++){
                        int co = sub.get(j);
                        courses[co]--;
                        if(courses[co]==0){
                            q.add(co);
                        }
                    }

                }
                
            }
        }
        for(int i=0;i<courses.length;i++){
            if(courses[i]!=0){
                return false;
            }
        }
        return true;


    }
}
